app/code/ZP/LoyaltyProgram/Controller/Adminhtml/Program/MassDelete.php

```diff
@@ -3,52 +3,135 @@
         Context $context,
-        private Filter $filter,
-        private CollectionFactory $collectionFactory,
-        private LoyaltyProgramResource $programResource
+        LoggerInterface $logger,
+        LoyaltyProgramRepositoryInterface $programRepository,
+        CustomerProgramManagement $customerProgramManagement,
+        Helper $helper,
+        ValidatorInterface $dataValidator,
+        RequestHelper $requestHelper,
+        Filter $filter,
+        CollectionFactory $collectionFactory,
+        DataPreparer $prepareData,
+        SalesRuleProgramsManagement $salesRuleProgramsManagement
     ) {
-        parent::__construct($context);
+        parent::__construct(
+            $context,
+            $logger,
+            $programRepository,
+            $customerProgramManagement,
+            $helper,
+            $dataValidator,
+            $requestHelper,
+            $filter,
+            $collectionFactory,
+            $prepareData
+        );
+
+        $this->salesRuleProgramsManagement = $salesRuleProgramsManagement;
     }
 
     public function execute()
     {
-        $collection = $this->filter->getCollection($this->collectionFactory->create());
-        $collectionSize = $collection->getSize();
+        try {
+            /** @var Redirect $resultRedirect */
+            $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
+            $this->checkProgramIds(self::DELETE);
 
-        /** @var \Magento\Backend\Model\View\Result\Redirect $resultRedirect */
-        $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
+            /** @var Collection $collection */
+            $collection = $this->filter->getCollection($this->collectionFactory->create());
 
-        try {
-            /** @var LoyaltyProgram $program */
-            foreach ($collection as $program) {
-                $this->programResource->delete($program);
-            }
+            if (!$collection->getSize()) {
+                $this->messageManager->addNoticeMessage(
+                    __('Selected program(s) has(have) been probably deleted already!')
+                );
+            } else {
+                $this->validateCollectionPrograms($collection);
+                if ($this->programsForAction) {
+                    $this->programsData = $this->programsForAction;
+                    $this->beforeDelete();
+                    $this->deletedProgramsCount = count($this->programsForAction);
+
+                    /**
+                     * @var int $programId
+                     * @var LoyaltyProgram $program
+                     */
+                    foreach ($collection->getItems() as $program) {
+                        $this->programRepository->delete($program);
+                    }
 
-            $this->messageManager->addSuccessMessage(__('A total of %1 program(s) have been deleted.', $collectionSize));
+                    $this->afterDelete();
```

Comment:
No we cant, because my before and after actions which are in controllers have action logic that is not applicable to our resource model. But I will try to remember this moment and we will discuss it on our next live meeting, and may in our discussion we will find the way how to do it even more functional.